src/models/model.cpp

```diff
@@ -249,32 +249,159 @@ int32_t Tokenizer::TokenToTokenId(const char* token) const {
+      }
+
+      Ort::ThrowOnError(Ort::api->UpdateROCMProviderOptions(&ort_provider_options, keys.data(), values.data(), keys.size()));
+      session_options.AppendExecutionProvider_ROCM(ort_provider_options);
+#if USE_DML
+    } else if (provider_options.name == "dml") {
+      if (!GetDmlInterface()) {
+        LUID device_luid{};
+        LUID* p_device_luid{};
+        for (const auto& [name, value] : provider_options.options) {
+          if (name == "luid") {
+            if (auto separator_position = value.find(":"); separator_position != std::string::npos) {
+              device_luid.HighPart = std::stol(value.substr(0, separator_position));
+              device_luid.LowPart = std::stol(value.substr(separator_position + 1));
+              p_device_luid = &device_luid;
+            }
+          }
+        }
+
+        InitDmlInterface(p_device_luid);
+      }
+
+      if (!disable_graph_capture) {
+        session_options.AddConfigEntry("ep.dml.enable_graph_capture", "1");
+        session_options.AddConfigEntry("ep.dml.disable_memory_arena", "1");
+      }
+
+      SetDmlProvider(session_options);
+
+      if (is_primary_session_options)
+        p_device = GetDeviceInterface(DeviceType::DML);  // We use a DML allocator for input/output caches, but other tensors will use CPU tensors
+#endif
+    } else if (provider_options.name == "qnn") {
+      session_options.AddConfigEntry("ep.share_ep_contexts", "1");
+      std::unordered_map<std::string, std::string> opts;
+      for (auto& option : provider_options.options) {
+        opts.emplace(option.first, option.second);
+      }
+
+      // TODO set device_type_ in a less hacky way.
+      // now, all QNN EP enable_htp_shared_memory_allocator option values had better be consistent...
+      // on the other hand, not sure if is_primary_session_options is the right thing to check here.
+      if (const auto opt_it = opts.find("enable_htp_shared_memory_allocator");
+          opt_it != opts.end() && opt_it->second == "1") {
+        p_device = GetDeviceInterface(DeviceType::QNN);
+      }
+
+      session_options.AppendExecutionProvider("QNN", opts);
+    } else if (provider_options.name == "webgpu") {
+      p_device = GetDeviceInterface(DeviceType::WEBGPU);
+      std::unordered_map<std::string, std::string> opts;
+      for (auto& option : provider_options.options) {
+        opts.emplace(option.first, option.second);
+      }
+      session_options.AppendExecutionProvider("WebGPU", opts);
+    } else
+      throw std::runtime_error("Unknown provider type: " + provider_options.name);
+  }
+  return p_device;
+}
+
+// Trivial ONNX model that just returns a single float constant. Used below to create an OrtSession that
+// lets us get a device Ort::Allocator for each device type. This is necessary because the Ort::Allocator
+// needs to persist and is valid for the lifetime of this OrtSession.
+static const uint8_t g_trivial_model[] = {
```

Comment:
One of the concerns with this trivial model approach is how it impacts QNN EP. The QNN execution provider will try to compile the model down to a context binary every time the model is loaded through the ort session. Given the triviality of the model, it should be not too bad for the binary generation. But my concern is that this may increase model load time as a result.

Reply:
Just tried it. This does impact the model load time a bit for QNN execution provider. Could we embed the context binary inside the ort-genai package for qnn?
Another thing I found is that the session options seem to enable verbose level logging, so my screen looks really scary when I run a model.

Reply:
ok, after doing some more testing. I think the model load time might also not be as bad as I thought it was. Seems fine. Sorry for the series of non-meaningful comments. :)

Reply:
Curious as to why there's any impact. I believe the trivial model has a Constant node and nothing else, which we convert to an initializer in the ORT Graph ctor leaving no actual nodes.